README.txt
episoder
========

Episoder is a tool to tell you about new episodes of your favourite TV shows. It
does so by parsing online TV episode guides.

Currently episoder can get data from epguides.com and thetvdb.com and tells you
whenever a new episode of a show listed on one of those sites is aired.

There is also a web-based version of episoder, called webisoder, which can be
used for free at webisoder.net.

Dependencies
------------

In order to use episoder, you will need python installed on your system. In
addition to the default distribution, the 'requests' and 'sqlalchemy' modules
are required.

Installation
------------

You can install episoder through pypi (using 'easy_install' or 'pip') or by
running 'setup.py' from the downloaded and extracted tarball.

Configuration
-------------

The configuration file at ~/.episoder contains your default settings for
episoder. After installing episoder, its manpage will help you with the
configuration.

Using episoder
--------------

To use episoder you will typically:

* Create a cron job to have your database rebuilt once a day by running
  `episoder update`
* Add `episoder` to your ~/.bashrc, ~/.bash_profile, /etc/bash.bashrc or
  /etc/profile if you want to see your upcoming shows every time you start a
  shell. Consult your shell's documentation to find out which file you want to
  use.
* Or run `episoder notify` at regular intervals to get your notifications by
  e-mail.


CHANGELOG
== v0.8.5 (28 Mar 2020) ==
 * Fix a small bug in setup.py

== v0.8.4 (28 Mar 2020) ==
 * Fix loading of shows without episodes from TVDB
 * Fix loading of episodes with invalid date (0000-00-00)

== v0.8.3 (27 Sep 2017) ==
 * Fixed a crash related to unexpected null values in TVDB JSON data
 * Use correct path for example config file in setup.py (fixes #7)

== v0.8.2 (7 Jul 2017) ==
 * Fixed a crash when rendering episodes without a title

== v0.8.1 (24 Jun 2017) ==
 * Handle failures to load TVDB API pages

== v0.8.0 (7 Jun 2017) ==
 * Switched from distutils to setuptools
 * We now use the new TVDB JSON API
 * Rewrote the epguides.com parser in pure Python
 * Rewrote large parts of the database code
 * Episoder should now work fully with Python 3
 * The update command can now be limited to a fixed number of shows
 * Updated man page
 * Thanks to Chris Spencer (chrisspen) for his contributions
 * Added option to only list active shows (chrisspen)
 * Added the number of episodes to the list of shows (chrisspen)
 * Added email notification feature via 'episoder notify' (chrisspen)

== v0.7.3 (7 Nov 2015) ==
 * Fixed a syntax error in the list of shows
 * Fixed season number parsing for epguides.com (thanks to jethrogb)

== v0.7.2 (26 Sep 2015) ==
 * Dropped unneeded dependency on BeautifulSoup
 * Fixed episode number parsing for epguides.com (thanks to jethrogb)
 * Improved python3 compatibility

== v0.7.1 (27 Apr 2014) ==
 * Accept unicode characters in prodnum fields

== v0.7.0 (14 Oct 2013) ==
 * Added support for TheTVDB.com
 * Removed support for tv.com
 * Changed command-line arguments. See the man page for details
 * Removed the ability to specify shows in the configuration file
 * Reorganized the command line options. Check the documentation for details
 * Updated the epguides.com parser to ignore trailer links in episode names
 * Fixed some more encoding issues in the console renderer
 * Changed license to GPLv3

== v0.6.5 (6 Oct 2010) ==
 * Updated the epguides.com parser to ignore links to trailers in episode names

== v0.6.4 (25 Jun 2010) ==
 * Filter javascript from tv.com pages because it confuses beautifulsoup 3.1

== v0.6.3 (23 Jun 2010) ==
 * The epguides parser now properly reads the season number for more shows

== v0.6.2 (08 Nov 2009) ==
 * Updated tv.com parser (show names need to be lower-case now!)
 * Added -S option to specify a show id for (forced) updates

== v0.6.1 (26 Aug 2009) ==
 * Episoder can now talk to any database supported by sqlalchemy (see README)
 * Tell awk to process utf8 (for epguides parser)
 * Catch (and properly log) some more HTTP exceptions

== v0.6.0 (8 Aug 2009) ==
 * Episoder now keeps a persistent database of shows and episodes
 * Added new options (-L, -a, -r, -E, -D) to manage the database
 * The parsers now know whether a show is still running or not
   * Active (currently running) shows are updated every two days
   * Inactive (ended) shows are updated every two weeks
   * The new -B command overrides those intervals and forces an update
 * Show names can no longer be overridden in the configuration file
 * Don't report an error when no sources are defined in the config file
 * The tv.com parser now also accepts urls containing '!', '+' or '%'
 * The -d option now also accepts the number of days instead of a fixed date
 * Documentation updated

== v0.5.3 (26 Jul 2009) ==
 * Added support for another epguides formatting style (used eg. for "Eureka")
 * Added option to suppress colors in console output
 * Added a filter to remove invalid characters from epguides content
 * Fixed some other parsing issues in the epguides parser
 * Improved compatibility with older python versions

== v0.5.2 (30 Jun 2009) ==
 * Reworked parts of the code to improve compatibility with older libraries
 * Fixed some encoding issues in the console renderer

== v0.5.1 (23 Jun 2009) ==
 * Thanks to Philip Douglass for his bug reports and patches
 * The epguides parser now properly closes all file handles and works with mawk
 * The tv.com parser would crash on missing production/season/episode numbers
 * The tv.com parser now accepts urls that contain a dash
 * Switched to using sqlalchemy to access the database
 * Properly implemented the user-agent configuration option
 * Removed old FAQ
 * Fixed some text encoding issues
 * Small fixes to the man page

== v0.5.0 (18 May 2009) ==
 * Rewrote the whole thing in python
 * Wrote a brand new tv.com parser
 * Small fixes in the epguides.com parser
 * Parsed data is now stored in an sqlite3 database

== v0.4.7 (24 Feb 2007) ==
 * Fixed a parsing issue introduced with 0.4.6

== v0.4.6 (24 Feb 2007) ==
 * Discovered yet another formatting possibility on epguides.com (used for
   instance for the show "Midsomer Murders") and adapted the parser accordingly
 * Slightly updated the man page

== v0.4.5 (23 Jan 2007) ==
 * Fixed some epguides parser issues

== v0.4.4 (19 Nov 2006) ==
 * Fixed the -i option when building the database (did the exact opposite of
   what it was supposed to do)

== v0.4.3 (24 Oct 2006) ==
 * We now actually honor the -i option when building the database
 * Added the possibility to override a show's name in the config file. This was needed for the new webisoder, a web-based episoder. Find out more about it at http://www.webisoder.net/
 * Oops, forgot to update episoder's version number in the last release...
 * Fixed a bug in the new epguides.com parser which would not drop episodes without a date

== v0.4.2 (21 Oct 2006) ==
 * Don't ignore the config file passed with -c anymore when building the db
 * Added parser for epguides.com (closes #38)

== v0.4.1 (20 Apr 2006) ==
 * Did some minor internal changes to get episoder to work on non-linux systems
 * Removed long options
 * Hotfix for the tv.com website changes

== v0.4.0 (3 Dec 2005) ==
 * Re-organized options: -vv is now -w, -rc is now -c (or --rc), -nd is gone (use --nodate), -sa is gone (use -s and --nodate instead), -i is gone (see below)

test/show.py
# episoder, https://code.ott.net/episoder
# -*- coding: utf8 -*-
#
# Copyright (C) 2004-2020 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from datetime import date, timedelta
from tempfile import mktemp
from unittest import TestCase, TestSuite, TestLoader
from os import unlink

from pyepisoder.episoder import Database
from pyepisoder.database import Episode, Show


class ShowTest(TestCase):

	def setUp(self):

		self.path = mktemp()
		self.db = Database(self.path)
		self.show = Show(u"A", url=u"a")
		self.show = self.db.add_show(self.show)

	def tearDown(self):

		unlink(self.path)

	def test_str_and_repr(self):

		self.assertEqual(str(self.show), "Show: A")
		self.assertEqual(repr(self.show), 'Show(u"A", u"a")')

	def test_equality(self):

		show = Show(u"A", url=u"b")
		self.assertNotEqual(show, self.show)

		show = Show(u"B", url=u"a")
		self.assertNotEqual(show, self.show)

		show = Show(u"A", url=u"a")
		self.assertEqual(show, self.show)

	def test_remove_episodes_before(self):

		now = date.today()
		then = now - timedelta(3)

		show2 = Show(u"B", url=u"b")
		show2 = self.db.add_show(show2)

		episode1 = Episode(u"e", 1, 1, now, u"x", 1)
		episode2 = Episode(u"e", 1, 2, then, u"x", 1)
		episode3 = Episode(u"e", 1, 3, now, u"x", 1)
		episode4 = Episode(u"e", 1, 4, then, u"x", 1)

		self.db.add_episode(episode1, self.show)
		self.db.add_episode(episode2, self.show)
		self.db.add_episode(episode3, show2)
		self.db.add_episode(episode4, show2)

		episodes = self.db.get_episodes(basedate = then, n_days=10)
		self.assertEqual(4, len(episodes))

		show2.remove_episodes_before(self.db, now)

		episodes = self.db.get_episodes(basedate = then, n_days=10)
		self.assertEqual(3, len(episodes))

		self.show.remove_episodes_before(self.db, now)

		episodes = self.db.get_episodes(basedate = then, n_days=10)
		self.assertEqual(2, len(episodes))


def test_suite():

	suite = TestSuite()
	loader = TestLoader()
	suite.addTests(loader.loadTestsFromTestCase(ShowTest))
	return suite


test/tvdb.py
# episoder, https://code.ott.net/episoder
# -*- coding: utf8 -*-
#
# Copyright (C) 2004-2020 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging

from datetime import date, datetime
from json import loads
from unittest import TestCase, TestSuite, TestLoader

import requests

from pyepisoder.database import Show
from pyepisoder.episoder import Database
from pyepisoder.sources import TVDB, TVDBNotLoggedInError, InvalidLoginError
from pyepisoder.sources import TVDBShowNotFoundError

from .util import MockResponse, MockArgs, LoggedRequest


class MockRequestHandler(object):

	def __init__(self):

		self.requests = []

	def load_fixture(self, name):

		with open("test/fixtures/tvdb_%s.json" % name, "rb") as file:
			data = file.read()

		return data

	def fixture(self, name, code):

		response = self.load_fixture(name)
		return MockResponse(response, "utf8", code)

	def get_search(self, url, headers, params):

		term = params.get("name")

		if term == "Frasier":
			return self.fixture("search_frasier", 200)
		elif term == "Friends":
			return self.fixture("search_friends", 200)

		return self.fixture("error_404", 404)

	def get_episodes(self, url, headers, params):

		id = int(url.split("/")[-2])
		p = params.get("page", 1)

		try:
			return self.fixture("test_show_%d_%d_eps" % (id,p), 200)
		except IOError:
			return self.fixture("error_404", 404)

	def get_show(self, url, headers, params):

		id = int(url.split("/")[-1])

		try:
			return self.fixture("test_show_%d" % id, 200)
		except IOError:
			return self.fixture("error_404", 404)

	def get(self, url, headers={}, params={}):

		req = LoggedRequest("GET", url, "", headers, params)
		self.requests.append(req)

		if url.startswith("https://api.thetvdb.com/search/series"):
			return self.get_search(url, headers, params)
		elif url.startswith("https://api.thetvdb.com/series"):
			if url.endswith("/episodes"):
				return self.get_episodes(url, headers, params)
			else:
				return self.get_show(url, headers, params)

		return MockResponse("{}", "utf8", 404)

	def post_login(self, body, headers):

		data = loads(body)
		key = data.get("apikey")

		if key == "fake-api-key":
			text = '{ "token": "fake-token" }'
			return MockResponse(text.encode("utf8"), "utf8", 200)

		text = '{"Error": "Not Authorized"}'
		return MockResponse(text.encode("utf8"), "utf8", 401)

	def post(self, url, body, headers = {}):

		req = LoggedRequest("POST", url, body, headers, {})
		self.requests.append(req)

		if url.startswith("https://api.thetvdb.com/login"):
			return self.post_login(body.decode("utf8"), headers)

		return MockResponse("{}", "utf8", 404)


class TVDBTest(TestCase):

	def setUp(self):

		logging.disable(logging.ERROR)

		self.db = Database("sqlite://")
		self.tvdb = TVDB()
		self.req = MockRequestHandler()
		self.args = MockArgs("fake-api-key", "episoder/test")

		self.__get_orig = requests.get
		self.__post_orig = requests.post
		requests.get = self.req.get
		requests.post = self.req.post

	def tearDown(self):

		requests.get = self.__get_orig
		requests.post = self.__post_orig

	def test_parser_name(self):

		self.assertEqual("thetvdb.com parser (ready)", str(self.tvdb))
		self.assertEqual("TVDB <TVDBOffline>", repr(self.tvdb))

		self.tvdb.login(MockArgs("fake-api-key"))
		self.assertEqual("thetvdb.com parser (authorized)",
								str(self.tvdb))
		self.assertEqual("TVDB <TVDBOnline>", repr(self.tvdb))

	def test_need_login(self):

		with self.assertRaises(TVDBNotLoggedInError):
			self.tvdb.lookup(u"Frasier", self.args)

		self.tvdb.login(MockArgs("fake-api-key"))
		self.tvdb.lookup(u"Frasier", self.args)

		self.tvdb.login(MockArgs("fake-api-key"))
		self.tvdb.lookup(u"Frasier", self.args)

	def test_login(self):

		self.tvdb.login(MockArgs("fake-api-key"))

		reqs = len(self.req.requests)
		self.assertTrue(reqs > 0)

		req = self.req.requests[-1]
		self.assertEqual(req.url, "https://api.thetvdb.com/login")
		self.assertEqual(req.method, "POST")
		self.assertEqual(req.body.decode("utf8"),
						'{"apikey": "fake-api-key"}')
		headers = req.headers
		self.assertEqual(headers.get("Content-type"),"application/json")

		self.tvdb.login(MockArgs("fake-api-key"))
		self.assertEqual(reqs, len(self.req.requests))

	def test_login_failure(self):

		with self.assertRaises(InvalidLoginError):
			self.tvdb.login(MockArgs("wrong-api-key"))

		with self.assertRaises(InvalidLoginError):
			self.tvdb.login(MockArgs("wrong-api-key"))

		with self.assertRaises(InvalidLoginError):
			self.tvdb.login(MockArgs("wrong-api-key"))

		self.tvdb.login(MockArgs("fake-api-key"))

	def test_search_no_hit(self):

		self.tvdb.login(MockArgs("fake-api-key"))

		with self.assertRaises(TVDBShowNotFoundError):
			self.tvdb.lookup("NoSuchShow", self.args)

	def test_search_single(self):

		self.tvdb.login(MockArgs("fake-api-key"))

		shows = list(self.tvdb.lookup("Frasier", self.args))

		req = self.req.requests[-1]
		self.assertEqual(req.url,
					"https://api.thetvdb.com/search/series")
		self.assertEqual(req.params, {"name": "Frasier"})
		self.assertEqual(req.method, "GET")
		self.assertEqual(req.body, "")

		content_type = req.headers.get("Content-type")
		self.assertEqual(content_type, "application/json")

		auth = req.headers.get("Authorization")
		self.assertEqual(auth, "Bearer fake-token")

		self.assertEqual(len(shows), 1)
		show = shows[0]
		self.assertEqual(show.name, "Frasier")
		self.assertEqual(show.url, "77811")

	def test_search_multiple(self):

		self.tvdb.login(MockArgs("fake-api-key"))

		shows = list(self.tvdb.lookup("Friends", self.args))

		self.assertEqual(len(shows), 3)
		self.assertEqual(shows[0].name, "Friends")
		self.assertEqual(shows[1].name, "Friends (1979)")
		self.assertEqual(shows[2].name, "Friends of Green Valley")

	def test_accept_url(self):

		self.assertTrue(TVDB.accept("123"))
		self.assertFalse(TVDB.accept("http://www.epguides.com/test"))

	def test_encoding_utf8(self):

		self.tvdb.login(self.args)

		show = self.db.add_show(Show(u"unnamed show", url=u"73739"))
		self.assertTrue(TVDB.accept(show.url))

		self.tvdb.parse(show, self.db, self.args)

		self.assertEqual("Lost", show.name)
		self.assertEqual(Show.ENDED, show.status)

		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 1)

		episode = episodes[0]
		self.assertEqual(episode.title, u"Exposé")

	def test_null_values(self):

		self.tvdb.login(self.args)
		show = self.db.add_show(Show(u"unnamed show", url=u"268156"))
		self.assertTrue(TVDB.accept(show.url))

		# this show has some null values that can confuse the parser
		self.tvdb.parse(show, self.db, self.args)
		self.assertEqual("Sense8", show.name)

	def test_parse(self):

		show = self.db.add_show(Show(u"unnamed show", url=u"260"))
		self.assertTrue(TVDB.accept(show.url))

		with self.assertRaises(TVDBNotLoggedInError):
			self.tvdb.parse(show, None, self.args)

		self.tvdb.login(self.args)
		self.tvdb.parse(show, self.db, self.args)

		req = self.req.requests[-2]
		self.assertEqual(req.url, "https://api.thetvdb.com/series/260")
		self.assertEqual(req.params, {})
		self.assertEqual(req.method, "GET")
		self.assertEqual(req.body, "")

		content_type = req.headers.get("Content-type")
		self.assertEqual(content_type, "application/json")

		auth = req.headers.get("Authorization")
		self.assertEqual(auth, "Bearer fake-token")

		req = self.req.requests[-1]
		self.assertEqual(req.url,
				"https://api.thetvdb.com/series/260/episodes")
		self.assertEqual(req.params, {"page": 1})
		self.assertEqual(req.method, "GET")
		self.assertEqual(req.body, "")

		content_type = req.headers.get("Content-type")
		self.assertEqual(content_type, "application/json")

		auth = req.headers.get("Authorization")
		self.assertEqual(auth, "Bearer fake-token")

		self.assertEqual(show.name, "test show")
		self.assertEqual(show.status, Show.RUNNING)

		timediff = datetime.now() - show.updated
		self.assertTrue(timediff.total_seconds() < 1)

		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 2)

		episode = episodes[0]
		self.assertEqual(episode.title, "Unnamed episode")
		self.assertEqual(episode.season, 0)
		self.assertEqual(episode.episode, 0)
		self.assertEqual(episode.airdate, date(1990, 1, 18))
		self.assertEqual(episode.prodnum, "UNK")
		self.assertEqual(episode.totalnum, 1)

		episode = episodes[1]
		self.assertEqual(episode.title, "The Good Son")
		self.assertEqual(episode.season, 1)
		self.assertEqual(episode.episode, 1)
		self.assertEqual(episode.airdate, date(1993, 9, 16))
		self.assertEqual(episode.totalnum, 2)

	def test_parse_paginated(self):

		show = self.db.add_show(Show(u"unnamed show", url=u"261"))

		self.tvdb.login(self.args)
		self.tvdb.parse(show, self.db, self.args)

		self.assertEqual(show.status, Show.ENDED)
		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 8)

		episode = episodes[0]
		self.assertEqual(episode.title, "First")

		episode = episodes[1]
		self.assertEqual(episode.title, "Second")

		episode = episodes[2]
		self.assertEqual(episode.title, "Third")

		episode = episodes[3]
		self.assertEqual(episode.title, "Fourth")

		episode = episodes[4]
		self.assertEqual(episode.title, "Fifth")

		episode = episodes[5]
		self.assertEqual(episode.title, "Sixth")

		episode = episodes[6]
		self.assertEqual(episode.title, "Seventh")

		episode = episodes[7]
		self.assertEqual(episode.title, "Eighth")

	def test_parse_invalid_show(self):

		self.tvdb.login(self.args)

		show = self.db.add_show(Show(u"test show", url=u"293"))

		with self.assertRaises(TVDBShowNotFoundError):
			self.tvdb.parse(show, None, self.args)

	def test_parse_show_with_invalid_data(self):

		self.tvdb.login(self.args)
		show = self.db.add_show(Show(u"unnamed show", url=u"262"))

		self.tvdb.parse(show, self.db, self.args)
		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 2)

	def test_parse_show_without_episodes(self):

		show = self.db.add_show(Show(u"unnamed show", url=u"263"))

		self.tvdb.login(self.args)
		self.tvdb.parse(show, self.db, self.args)

		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 0)

	def test_parse_show_with_data_0000_00_00(self):

		show = self.db.add_show(Show(u"unnamed show", url=u"75397"))

		self.tvdb.login(self.args)
		self.tvdb.parse(show, self.db, self.args)
		episodes = self.db.get_episodes(date(1988,1,1), 99999)
		self.assertEqual(len(episodes), 1)

	def test_user_agent(self):

		show = self.db.add_show(Show(u"unnamed show", url=u"262"))

		self.tvdb.login(self.args)
		self.tvdb.lookup("Frasier", self.args)
		self.tvdb.parse(show, self.db, self.args)

		self.assertEqual(len(self.req.requests), 4)

		req = self.req.requests[0]
		headers = req.headers
		self.assertEqual(headers.get("User-Agent"), "episoder/test")

		req = self.req.requests[1]
		headers = req.headers
		self.assertEqual(headers.get("User-Agent"), "episoder/test")

		req = self.req.requests[2]
		headers = req.headers
		self.assertEqual(headers.get("User-Agent"), "episoder/test")

		req = self.req.requests[3]
		headers = req.headers
		self.assertEqual(headers.get("User-Agent"), "episoder/test")


def test_suite():

	suite = TestSuite()
	loader = TestLoader()
	suite.addTests(loader.loadTestsFromTestCase(TVDBTest))
	return suite


test/util.py
# episoder, https://code.ott.net/episoder
# -*- coding: utf8 -*-
#
# Copyright (C) 2004-2020 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import json


class RawWrapper(object):

	def __init__(self, text):

		self._text = text

	def read(self):

		return self._text


class MockResponse(object):

	def __init__(self, data, encoding, status=200):

		self.raw = RawWrapper(data)
		self.encoding = encoding
		self.status_code = status

	def _get_text(self):

		text = self.raw.read()
		return text.decode(self.encoding, "replace")

	def json(self):

		return json.loads(self.text)

	text = property(_get_text)


class MockArgs(dict):

	def __init__(self, tvdb_key, agent=None):

		self.tvdb_key = tvdb_key
		self.agent = agent


class LoggedRequest(object):

	def __init__(self, method, url, body, headers, params):

		self.method = method
		self.url = url
		self.body = body
		self.headers = headers
		self.params = params


pyepisoder/sources.py
# episoder, https://code.ott.net/episoder
#
# Copyright (C) 2004-2020 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging

from re import search, match
from json import dumps
from datetime import datetime

import requests

from .database import Episode, Show


class ParserSelector(object):

	class _Inner(object):

		def __init__(self):

			self._tvdb = TVDB()
			self._epguides = Epguides()
			self._tvcom = TVCom()

		def parser_for(self, url):

			for parser in [self._tvdb, self._epguides, self._tvcom]:
				if parser.accept(url):
					return parser

			return None

	_instance = None

	@staticmethod
	def instance():

		if not ParserSelector._instance:
			ParserSelector._instance = ParserSelector._Inner()

		return ParserSelector._instance


class InvalidLoginError(Exception):

	pass


class TVDBShowNotFoundError(Exception):

	pass


class TVDBNotLoggedInError(Exception):

	pass


class TVDBOffline(object):

	def __init__(self, tvdb):

		self._tvdb = tvdb
		self._log = logging.getLogger("TVDB (offline)")

	def __str__(self):

		return "thetvdb.com parser (ready)"

	def __repr__(self):

		return "<TVDBOffline>"

	def _post_login(self, data, user_agent):

		url = "https://api.thetvdb.com/login"
		head = {"Content-type": "application/json",
			"User-Agent": user_agent}
		body = dumps(data).encode("utf8")
		response = requests.post(url, body, headers=head)
		data = response.json()

		if response.status_code == 401:
			raise InvalidLoginError(data.get("Error"))

		self._log.info("Successful login")
		return data.get("token")

	def lookup(self, _, __):

		raise TVDBNotLoggedInError()

	def login(self, args):

		body = {"apikey": args.tvdb_key}
		self.token = self._post_login(body, args.agent)

	def parse(self, _, __, ___):

		raise TVDBNotLoggedInError()

	def _set_token(self, token):

		self._tvdb.change(TVDBOnline(token))

	token = property(None, _set_token)


class TVDBOnline(object):

	def __init__(self, token):

		self._token = token
		self._log = logging.getLogger("TVDB (online)")

	def __str__(self):

		return "thetvdb.com parser (authorized)"

	def __repr__(self):

		return "<TVDBOnline>"

	def _get(self, url, params, agent):

		url = "https://api.thetvdb.com/%s" % url
		head = {"Content-type": "application/json",
			"User-Agent": agent,
			"Authorization": "Bearer %s" % self._token}
		response = requests.get(url, headers=head, params=params)
		data = response.json()

		if response.status_code == 404:
			raise TVDBShowNotFoundError(data.get("Error"))

		return data

	def _get_episodes(self, show, page, agent):

		show_id = int(show.url)
		opts = {"page": page}
		result = self._get("series/%d/episodes" % show_id, opts, agent)
		return (result.get("data"), result.get("links"))

	def lookup(self, term, agent):

		def mkshow(entry):

			name = entry.get("seriesName")
			url = str(entry.get("id")).encode("utf8").decode("utf8")
			return Show(name, url=url)

		matches = self._get("search/series", {"name": term}, agent)
		return map(mkshow, matches.get("data"))

	def login(self, args):

		pass

	def _fetch_episodes(self, show, page, agent):

		def airdate(row):

			aired = row.get("firstAired")

			if aired == "0000-00-00":
				return datetime(1970, 1, 1)

			return datetime.strptime(aired, "%Y-%m-%d").date()

		def mkepisode(row):

			num = int(row.get("airedEpisodeNumber", "0") or 0)
			name = row.get("episodeName") or u"Unnamed episode"
			season = int(row.get("airedSeason", "0") or 0)
			aired = airdate(row)
			pnum = u"UNK"

			self._log.debug("Found episode %s", name)
			return Episode(name, season, num, aired, pnum, 0)

		def isvalid(row):

			return row.get("firstAired") not in [None, ""]

		(data, links) = self._get_episodes(show, page, agent)
		valid = filter(isvalid, data)
		episodes = [mkepisode(row) for row in valid]

		# handle pagination
		next_ = links.get("next") or 0
		if next_ > page:
			try:
				more = self._fetch_episodes(show, next_, agent)
				episodes.extend(more)
			except TVDBShowNotFoundError:
				msg = "Error parsing %s: failed to load page %d"
				self._log.error(msg, show.name, next_)

		return episodes

	def parse(self, show, db, user_agent):

		result = self._get("series/%d" % int(show.url), {}, user_agent)
		data = result.get("data")

		# update show data
		show.name = data.get("seriesName", show.name)
		show.updated = datetime.now()

		self._log.debug("Updating show '%s'", show.name)

		if data.get("status") == "Continuing":
			show.status = Show.RUNNING
		else:
			show.status = Show.ENDED

		# load episodes
		try:
			episodes = self._fetch_episodes(show, 1, user_agent)
		except TVDBShowNotFoundError:
			#raise
			return

		for (idx, episode) in enumerate(sorted(episodes)):

			episode.totalnum = idx + 1
			db.add_episode(episode, show)

		db.commit()


class TVDB(object):

	def __init__(self):

		self._state = TVDBOffline(self)

	def __str__(self):

		return str(self._state)

	def __repr__(self):

		return "TVDB %s" % repr(self._state)

	def login(self, args):

		self._state.login(args)

	def lookup(self, text, args):

		return self._state.lookup(text, args.agent)

	def parse(self, show, db, args):

		return self._state.parse(show, db, args.agent)

	def change(self, state):

		self._state = state

	@staticmethod
	def accept(url):

		return url.isdigit()


class Epguides(object):

	def __init__(self):

		self.logger = logging.getLogger("Epguides")

	def __str__(self):

		return "epguides.com parser"

	def __repr__(self):

		return "Epguides()"

	@staticmethod
	def accept(url):

		return "epguides.com/" in url

	def login(self, args):

		pass

	def guess_encoding(self, response):

		raw = response.raw.read()
		text = raw.decode("iso-8859-1")

		if "charset=iso-8859-1" in text:
			return "iso-8859-1"

		return "utf8"

	def parse(self, show, db, args):

		headers = {"User-Agent": args.agent}
		response = requests.get(show.url, headers=headers)
		response.encoding = self.guess_encoding(response)

		for line in response.text.split("\n"):
			self._parse_line(line, show, db)

		show.updated = datetime.now()
		db.commit()

	def _parse_line(self, line, show, db):

		# Name of the show
		res = search(r"<title>(.*)</title>", line)
		if res:
			title = res.groups()[0]
			show.name = title.split(" (a ")[0]

		# Current status (running / ended)
		res = search(r'<span class="status">(.*)</span>', line)
		if res:
			text = res.groups()[0]
			if "current" in text:
				show.status = Show.RUNNING
			else:
				show.status = Show.ENDED
		else:
			if search(r"aired.*to.*[\d+]", line):
				show.status = Show.ENDED

		# Known formatting supported by this fine regex:
		# 4.     1-4            19 Jun 02  <a [..]>title</a>
		#   1.  19- 1   01-01    5 Jan 88  <a [..]>title</a>
		# 23     3-05           27/Mar/98  <a [..]>title</a>
		# 65.   17-10           23 Apr 05  <a [..]>title</a>
		# 101.   5-15           09 May 09  <a [..]>title</a>
		# 254.    - 5  05-254   15 Jan 92  <a [..]>title</a>
		res = search(r"^ *(\d+)\.? +(\d*)- ?(\d+) +([a-zA-Z0-9-]*)"\
		r" +(\d{1,2}[ /][A-Z][a-z]{2}[ /]\d{2}) *<a.*>(.*)</a>", line)

		if res:
			fields = res.groups()
			(total, season, epnum, prodnum, day, title) = fields

			day = day.replace("/", " ")
			airtime = datetime.strptime(day, "%d %b %y")

			self.logger.debug("Found episode %s", title)
			db.add_episode(Episode(title, season or 0, epnum,
					airtime.date(), prodnum, total), show)


class TVCom(object):

	def __init__(self):

		self._log = logging.getLogger("TVCom")

	def __str__(self):

		return "dummy tv.com parser to detect old urls"

	def __repr__(self):

		return "TVCom()"

	@staticmethod
	def accept(url):

		exp = "http://(www.)?tv.com/.*"
		return match(exp, url)

	def parse(self, source, _, __):

		self._log.error("The url %s is no longer supported", source.url)

	def login(self):

		pass
